File: sieve_primes.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

struct linkNode {
    int value;
    struct linkNode *next;
};

void printAllNodes(struct linkNode *start)
{
  struct linkNode *current = start;
  while(1) {
    if(current->next == NULL) {
      break;
    } else {
      printf("%d\n", current->value);
      current = current->next;
    }
  }
}
long countAllNodes(struct linkNode *start)
{
  long num = 0;
  struct linkNode *current = start;
  while(1) {
    if(current->next == NULL) {
      break;
    } else {
      current = current->next;
      num++;
    }
  }
  return num;
}

void freeAllNodes(struct linkNode *start)
{
  struct linkNode *current = start;
  while(1) {
    if(current->next == NULL) {
      break;
    } else {
      struct linkNode *old = current;
      current = current->next;
      free(old);
    }
  }
}

// 2 3 5 7....
int main(int argc, char* argv[])
{
  if(argc < 2) {
    printf("arg error");
    return 1;
  }
  long max = atoi(argv[1]);
  int end = ceil(sqrt(max));
  
  printf("Max = %ld End = %d \n", max, end);
  struct linkNode *start;
  struct linkNode *current;
  start = (struct linkNode *)malloc(sizeof(struct linkNode));
  current = start;

  current->next = (struct linkNode *)malloc(sizeof(struct linkNode));
  current->next->next = NULL;
  current->value = 2;
  current = current->next;

  for(int i = 2;i < max;i += 2) {
    current->next = (struct linkNode *)malloc(sizeof(struct linkNode));
    current->next->next = NULL;
    current->value = i + 1;
    current = current->next;
  }
  current = start;

  while(1) {
    if(current->next == NULL || current->next->value == 0) {
      break;
    } 
    int num = current->value;
    // printf("N = %d\n", num);

    struct linkNode *seek = current;
    while(1) {
      // printf("seek.value = %d\n", seek->value);
      if(seek->next == NULL || seek->next->value == 0) {
        break;
      }
      if(seek->next->value % num == 0) {
          // printf("seek.next.value = %d\n", seek->next->value);
          struct linkNode *old = seek->next;
          seek->next = seek->next->next;
          free(old);
      }
      seek = seek->next;
      if(seek == NULL) {
        break;
      }
    }
    current = current->next;
  }
  printf("%ld primes\n", countAllNodes(start));

  freeAllNodes(start);
  return 0;
}

File: sieve_primes.go
package main

import (
  "fmt"
  "math"
  "os"
  "strconv"
)

type LinkNode struct {
    value int
    next *LinkNode
}

func CountAllNodes(start *LinkNode) int{
  num := 0
  current := start;
  for {
    if(current.next == nil) {
      break
    } else {
      current = current.next
      num += 1
    }
  }
  return num
}

// 2 3 5 7....
func main() {
  if(len(os.Args) < 2) {
    fmt.Printf("arg error")
    os.Exit(1)
  }
  
  max64, err := strconv.ParseInt(os.Args[1], 10, 64)
  if err != nil {
    panic("parse error")
  }
  max := int(max64)
  end := math.Ceil(math.Sqrt(float64(max)))

  fmt.Printf("Max = %d End = %d \n", max, int(end))

  start := new(LinkNode)
  var current *LinkNode = start

  (*current).next = new(LinkNode)
  (*current).next.next = nil
  (*current).value = 2
  current = (*current).next

  for i := 2;i < max;i += 2 {
    (*current).next = new(LinkNode)
    (*current).next.next = nil
    (*current).value = i + 1
    current = (*current).next
  }
  current = start

  for {
    if current.next == nil || current.next.value == 0 {
      break
    } 
    num := current.value

    seek := current
    for {
      if seek.next == nil || seek.next.value == 0 {
        break
      }
      if seek.next.value % num == 0 {
          seek.next = seek.next.next
      }
      seek = seek.next
      if seek == nil {
        break
      }
    }
    current = current.next
  }
  fmt.Printf("%d primes\n", CountAllNodes(start))
}
